adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+	if openRTBRequest.Regs != nil && openRTBRequest.Regs.COPPA >= 0 {
+		var regs regs
+		regs.Coppa = int32(openRTBRequest.Regs.COPPA)
+		request.Regs = regs
+	}
+}
+
+// get geo information for HuaweiAds request
+func getHuaweiAdsReqGeoInfo(request *huaweiAdsRequest, openRTBRequest *openrtb2.BidRequest) {
+	if openRTBRequest.Device != nil && openRTBRequest.Device.Geo != nil {
+		var geo geo
+		geo.Lon = float32(openRTBRequest.Device.Geo.Lon)
+		geo.Lat = float32(openRTBRequest.Device.Geo.Lat)
+		geo.Accuracy = int32(openRTBRequest.Device.Geo.Accuracy)
+		geo.Lastfix = int32(openRTBRequest.Device.Geo.LastFix)
+		request.Geo = geo
+	}
+}
+
+func unmarshalExtImpHuaweiAds(openRTBImp *openrtb2.Imp) (*openrtb_ext.ExtImpHuaweiAds, error) {
+	var bidderExt adapters.ExtImpBidder
+	var huaweiAdsImpExt openrtb_ext.ExtImpHuaweiAds
+	if err := json.Unmarshal(openRTBImp.Ext, &bidderExt); err != nil {
+		return nil, errors.New("Unmarshal: openRTBImp.Ext -> bidderExt failed")
+	}
+	if err := json.Unmarshal(bidderExt.Bidder, &huaweiAdsImpExt); err != nil {
+		return nil, errors.New("Unmarshal: bidderExt.Bidder -> huaweiAdsImpExt failed")
+	}
+	if huaweiAdsImpExt.SlotId == "" {
+		return nil, errors.New("ExtImpHuaweiAds: slotid is empty.")
+	}
+	if huaweiAdsImpExt.Adtype == "" {
+		return nil, errors.New("ExtImpHuaweiAds: adtype is empty.")
+	}
+	if huaweiAdsImpExt.PublisherId == "" {
+		return nil, errors.New("ExtHuaweiAds: publisherid is empty.")
+	}
+	if huaweiAdsImpExt.SignKey == "" {
+		return nil, errors.New("ExtHuaweiAds: signkey is empty.")
+	}
+	if huaweiAdsImpExt.KeyId == "" {
+		return nil, errors.New("ExtImpHuaweiAds: keyid is empty.")
+	}
+	return &huaweiAdsImpExt, nil
+}
+
+// check response status code
+func checkRespStatusCode(response *adapters.ResponseData) error {
+	if response.StatusCode == http.StatusNoContent {
+		return nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: [ %d ]", response.StatusCode),
+		}
+	}
+
+	if response.StatusCode == http.StatusServiceUnavailable {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Something went wrong, please contact your Account Manager. Status Code: [ %d ] ", response.StatusCode),
+		}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: [ %d ]. Run with request.debug = 1 for more info", response.StatusCode),
+		}
+	}
+
+	if response.Body == nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("bidderRawResponse body is empty"),
+		}
+	}
+	return nil
+}
+
+// check HuaweiAds response retcode
+func checkHuaweiAdsResponseRetcode(response huaweiAdsResponse) error {
+	if response.Retcode == 200 || response.Retcode == 204 || response.Retcode == 206 {
+		return nil
+	}
+
+	if (response.Retcode < 600 && response.Retcode >= 400) || (response.Retcode < 300 && response.Retcode > 200) {
```

Comment:
Please add test coverage for this condition.

Reply:
added bad_response_retcode_210.json bad_response_retcode_408.json bad_response_retcode_500.json